### PR TITLE
Adds scrape timeout property to ServiceMonitor endpoint

### PR DESCRIPTION
**Pull request by tao-lu**
In reference to #536 .

### CODE REVIEW

**Comment by brancz** [comment]
Generally I think this is something we want and the implementation also looks ok. Could you just run `make generate` to regenerate the documentation? (the API docs are generated from the struct code definitions)

**Comment by tao-lu** [comment]
Hmm, I'm getting the following output when I try to run `make generate`:

```C02T73JQFVH4-lm:prometheus-operator taol$ make generate
docker build -f scripts/jenkins/jsonnet/Dockerfile -t po-jsonnet .
Sending build context to Docker daemon  47.93MB
Step 1/8 : FROM golang:1.8-stretch
 ---> 85ce7a642216
Step 2/8 : ADD https://github.com/google/jsonnet/archive/v0.9.4.tar.gz /tmp
Downloading [==================================================>]  556.7kB/556.7kB
 ---> Using cache
 ---> 6d30b3facb57
Step 3/8 : RUN apt-get update -y && apt-get install -y g++ make git python-pip
 ---> Using cache
 ---> 957dcb4d6388
Step 4/8 : RUN cd /tmp && tar xvfz v0.9.4.tar.gz && cd jsonnet-0.9.4 && make && mv jsonnet /usr/local/bin && cd / && rm -rf /tmp/jsonnet-0.9.4 /tmp/jsonnet-0.9.4.tar.gz
 ---> Running in 260db9da144c
tar (child): v0.9.4.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
The command '/bin/sh -c cd /tmp && tar xvfz v0.9.4.tar.gz && cd jsonnet-0.9.4 && make && mv jsonnet /usr/local/bin && cd / && rm -rf /tmp/jsonnet-0.9.4 /tmp/jsonnet-0.9.4.tar.gz' returned a non-zero code: 2
make: *** [jsonnet-docker] Error 2```

Is this familiar to you?

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -153,6 +153,9 @@ func generateServiceMonitorConfig(version semver.Version, m *v1alpha1.ServiceMon
 	if ep.Interval != "" {
 		cfg = append(cfg, yaml.MapItem{Key: "scrape_interval", Value: ep.Interval})
 	}
+	if ep.Timeout != "" {
+		cfg = append(cfg, yaml.MapItem(Key: "scrape_timeout", Value: ep.Timeout})
```
Comment (brancz) [inline pkg/prometheus/promcfg.go]: `yaml.MapItem(` -> `yaml.MapItem{`

Comment (tao-lu) [inline pkg/prometheus/promcfg.go]: 👍

**Comment by tao-lu** [comment]
Okay, no problem; I just changed that line of the Dockerfile to `wget` the archive instead.

**Comment by brancz** [comment]
@coreosbot test this please

**Comment by brancz** [comment]
We're having some trouble with our test infrastructure. I'll re-run the tests when that's resolved.

File: Documentation/api.md [hunk 2]
```diff
@@ -92,6 +92,7 @@ Endpoint defines a scrapeable endpoint serving Prometheus metrics.
 | path | HTTP path to scrape for metrics. | string | false |
 | scheme | HTTP scheme to use for scraping. | string | false |
 | interval | Interval at which metrics should be scraped | string | false |
+| timeout | Timeout after which the scrape is ended | string | false |
```
Comment (gmcquillan) [inline Documentation/api.md]: Can we call this `scrapeTimeout` to match the attribute name in the prometheus config?

File: pkg/client/monitoring/v1alpha1/types.go [hunk 3]
```diff
@@ -200,6 +200,8 @@ type Endpoint struct {
 	Scheme string `json:"scheme,omitempty"`
 	// Interval at which metrics should be scraped
 	Interval string `json:"interval,omitempty"`
+	// Timeout after which the scrape is ended
+	Timeout string `json:"timeout,omitempty"`
```
Comment (gmcquillan) [inline pkg/client/monitoring/v1alpha1/types.go]: Again `scrapeTimeout` here?

File: Documentation/api.md (hunk 2 above)
Comment (tao-lu) [inline Documentation/api.md]: Sure thing.

**Comment by gmcquillan** [comment]
I'm 👍

**Comment by brancz** [comment]
Code lgtm, just want to give the e2e tests a spin, but unfortunately we're still having some trouble. Sorry for the inconvenience!

**Comment by gmcquillan** [comment]
@brancz Would it be at all possible to kick it off locally and merge and cut a release if the e2e's pass?

**Comment by brancz** [comment]
Typically yes, but we are still in a discussion and have more or less concluded that we will first need to get #551 done before we cut a new release, otherwise we have to do two migrations. So while I can kick off the tests locally and even merge here if they pass I won't be able to cut a release immediately, if you are ok with possibly handling the inconsistencies you could build from master then.

**Comment by brancz** [comment]
I ran the e2e tests locally and can confirm that they are passing. I'll merge here, but due to the reason above I can't say yet when the next release will go out.

**Comment by gmcquillan** [comment]
Thanks!

**Comment by gmcquillan** [comment]
I'm down with running an image from master. Can you push one up to quay?

**Pull request merged**